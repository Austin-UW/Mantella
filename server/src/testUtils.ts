import { ProjectFieldsFragment } from './types'
import { MutationResult } from 'react-apollo'
import { ApolloServer } from 'apollo-server-express'
import { queries } from './graphql/queries/queries'
import { mutations } from './graphql/mutations/mutations'
import path from 'path'
import fs from 'fs'
import mongoose from 'mongoose'

// Load models since we will not be instantiating our express server.
mongoose.connect(
  process.env.DB_CONNECT ||
    'mongodb+srv://conrad:cokay101@kanbanbrawn-41gbz.mongodb.net/kanban?retryWrites=true',
  { useNewUrlParser: true }
)

mongoose.set('useFindAndModify', false)
mongoose.set('useCreateIndex', true)

const { createTestClient } = require('apollo-server-testing')

const resolvers: any = {
  Query: queries,
  Mutation: mutations
}

const typeDefs: any = [
  fs.readFileSync(path.join(__dirname, './graphql/schema.graphql'), 'utf8')
]

const gqlServer = new ApolloServer({
  typeDefs,
  resolvers,
  tracing: false,
  context: ({ req, res }) => {
    return {
      req,
      res,
      user: {
        projects: ['5d018bc58104190fe6e2ac2c'],
        _id: '5d018bc58104190fe6e2ac2d',
        password:
          '$2a$10$LL5wFKVW./hWctu8lypf/.geDiezQhQG6rYdbI7fIsVd0wfzYJwha',
        email: 'conradkay123@gmjail.com',
        username: 'kachow',
        profileImg:
          'https://mb.cision.com/Public/12278/2797280/879bd164c711a736_800x800ar.png',
        __v: 0
      }
    }
  }
})

export const test = createTestClient(gqlServer)

export const gqlReq = async <Q, V>(args: {
  query: any
  variables: V
  isMutation: boolean
  onError?: (err: any) => void
}): Promise<MutationResult<Q>> => {
  const result = args.isMutation
    ? await test.mutate({ mutation: args.query, variables: args.variables })
    : await test.query({ query: args.query, variables: args.variables })

  return result
}

export const projToSnapshot = (project: ProjectFieldsFragment) => {
  return {
    ...project,
    id: 'PROJECT_ID',
    columns: project.columns.map((column, i) => {
      return {
        ...column,
        id: `COL_ID ${i}`,
        taskIds: column.taskIds.map((taskId, e) => `TASK_ID ${e}`)
      }
    }),
    columnIds: project.columnOrder.map((columnId, i) => `COL_ID ${i}`),
    ownerId: 'OWNER_ID',
    swimlanes: project.swimlanes.map((swimlane, i) => {
      return {
        ...swimlane,
        id: `SWIM_ID ${i}`,
        taskIds: swimlane.taskIds.map((taskId, e) => `TASK_ID ${e}`)
      }
    }),
    tags: project.tags!.map((tag, i) => {
      return { ...tag, id: `TAG_ID ${i}` }
    }),
    tasks: project.tasks.map((task, i) => {
      return {
        ...task,
        id: `TASK_ID ${i}`,
        subTasks: task.subTasks.map((subTask, e) => {
          return { ...subTask, id: `SUBTASK_ID ${e}` }
        }),
        comments: task.comments.map((comment, e) => {
          return { ...comment, id: `COMMENT_ID ${e}` }
        })
      }
    }),
    users: project.users!.map((user, i) => {
      return { id: `USER_ID ${i}` }
    })
  }
}
