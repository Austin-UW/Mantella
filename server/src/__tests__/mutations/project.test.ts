import {
  EditProjectMutationVariables,
  DeleteProjectMutation,
  DeleteProjectMutationVariables,
  CreateProjectMutation,
  CreateProjectMutationVariables,
  EditProjectMutation
} from './../../../../client/src/graphql/types'
import { gqlReq, projToSnapshot } from '../../testUtils'

import {
  GQL_CREATE_PROJECT,
  GQL_EDIT_PROJECT,
  GQL_DELETE_PROJECT
} from '../../../../client/src/graphql/mutations/project'

describe('Project Mutations', () => {
  test('createProject and editProject and deleteProject', async () => {
    /* createProject */
    const { data: createData } = await gqlReq<
      CreateProjectMutation,
      CreateProjectMutationVariables
    >({
      query: GQL_CREATE_PROJECT,
      variables: { name: 'project name' },
      isMutation: true
    })

    expect(createData).toBeTruthy()
    expect(createData!.createProject!.name).toBe('project name')

    expect(projToSnapshot(createData!.createProject!)).toMatchSnapshot()

    /* editProject */

    const { data: editData } = await gqlReq<
      EditProjectMutation,
      EditProjectMutationVariables
    >({
      query: GQL_EDIT_PROJECT,
      isMutation: true,
      variables: {
        id: createData!.createProject!.id,
        newProj: { name: 'new name lol', columnIds: ['real smooth'] }
      }
    })

    expect(editData).toBeTruthy()

    expect(editData!.editProject!.name).toBe('new name lol')
    expect(editData!.editProject!.columnOrder).toEqual(['real smooth'])
    expect(editData!.editProject!.id).toBe(createData!.createProject!.id)
    expect(projToSnapshot(editData!.editProject!)).toMatchSnapshot()

    /* deleteProject */

    const { data: deleteData } = await gqlReq<
      DeleteProjectMutation,
      DeleteProjectMutationVariables
    >({
      isMutation: true,
      query: GQL_DELETE_PROJECT,
      variables: {
        id: editData!.editProject!.id
      }
    })

    expect(deleteData!.deleteProject!.id).toBe(editData!.editProject!.id)
  })
})
