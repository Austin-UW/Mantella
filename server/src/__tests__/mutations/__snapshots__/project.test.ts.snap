// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Mutations createProject and editProject and deleteProject 1`] = `
Object {
  "columnIds": Array [
    "COL_ID 0",
  ],
  "columns": Array [
    Object {
      "id": "COL_ID 0",
      "isCompletedColumn": false,
      "name": "Todo",
      "taskIds": Array [],
      "taskLimit": 0,
    },
  ],
  "id": "PROJECT_ID",
  "name": "project name",
  "ownerId": "OWNER_ID",
  "swimlanes": Array [],
  "tags": Array [],
  "tasks": Array [],
  "users": Array [],
}
`;

exports[`Project Mutations createProject and editProject and deleteProject 2`] = `
Object {
  "columnIds": Array [
    "COL_ID 0",
  ],
  "columns": Array [
    Object {
      "id": "COL_ID 0",
      "isCompletedColumn": false,
      "name": "Todo",
      "taskIds": Array [],
      "taskLimit": 0,
    },
  ],
  "id": "PROJECT_ID",
  "name": "new name lol",
  "ownerId": "OWNER_ID",
  "swimlanes": Array [],
  "tags": Array [],
  "tasks": Array [],
  "users": Array [],
}
`;
